src/DrDotnet.Profilers/profilers/merged_call_stacks_profiler.rs

```diff
@@ -0,0 +1,254 @@
+            let method_ids_ptr_c = &method_ids as *const Vec<ffi::FunctionID> as *mut std::ffi::c_void;
+            
+            let _ =  pinfo.do_stack_snapshot(
+                managed_thread_id, 
+                crate::utils::stack_snapshot_callback, 
+                ffi::COR_PRF_SNAPSHOT_INFO::COR_PRF_SNAPSHOT_DEFAULT, 
+                method_ids_ptr_c, 
+                std::ptr::null(), 0);    
+
+            let mut stack_trace = Vec::<StackFrame>::new();
+            
+            for method_id in method_ids.iter().rev() {
+                let frame = StackFrame {
+                    kind: if *method_id == 0 { StackFrameType::Native } else { StackFrameType::Managed },
+                    fct_id: *method_id,
+                    // display: None
+                };
+                stack_trace.push(frame);
+            }
+            
+            if stack_trace.is_empty() { continue }
+            
+            merged_stack.add_stack(managed_thread_id, stack_trace, None);
+        }
+    }
+}
+
+impl CorProfilerCallback for MergedCallStacksProfiler {}
+
+impl CorProfilerCallback2 for MergedCallStacksProfiler {}
+
+impl CorProfilerCallback3 for MergedCallStacksProfiler {
+    fn initialize_for_attach(&mut self, profiler_info: ClrProfilerInfo, client_data: *const std::os::raw::c_void, client_data_length: u32) -> Result<(), ffi::HRESULT> {
+        self.init(ffi::COR_PRF_MONITOR::COR_PRF_ENABLE_STACK_SNAPSHOT, None, profiler_info, client_data, client_data_length)
+    }
+
+    fn profiler_attach_complete(&mut self) -> Result<(), ffi::HRESULT> {
+        
+        let profiler_info = self.clr().clone();
+        let merged_stack = self.merged_stack.clone();
+
+        let thread_handle = std::thread::spawn(move || {
+            
+            // https://github.com/dotnet/runtime/issues/37586#issuecomment-641114483
+            if profiler_info.suspend_runtime().is_ok() {
+                
+                let k = merged_stack.lock().unwrap();
+                MergedCallStacksProfiler::build_callstacks(profiler_info.clone(), k);
+                
+                if profiler_info.resume_runtime().is_err() {
+                    error!("Can't resume runtime!");
+                }
+            } else {
+                error!("Can't suspend runtime!");
+            }
+        });
+        
+        if thread_handle.join().is_err() {
+            error!("Can't wait for the thread to finish!");
+        }
+        
+        self.clr().request_profiler_detach(3000)
+    }
+
+    fn profiler_detach_succeeded(&mut self) -> Result<(), ffi::HRESULT> {
+        let mut report = self.session_info.create_report("summary.md".to_owned());
+        report.write_line(format!("# Merged Callstacks"));
+        
+        debug!("Hello");
+        let merged_stack = Arc::try_unwrap(self.merged_stack.clone()).unwrap().into_inner().unwrap();
```

Comment:
I get a crash on either the first or the second unwrap here

Reply:
yeah I simplified this part, but there are still plenty of other issues 🥲